python/docs/source/user_guide/sql/python_udtf.rst

```diff
@@ -0,0 +1,140 @@
+..  Licensed to the Apache Software Foundation (ASF) under one
+    or more contributor license agreements.  See the NOTICE file
+    distributed with this work for additional information
+    regarding copyright ownership.  The ASF licenses this file
+    to you under the Apache License, Version 2.0 (the
+    "License"); you may not use this file except in compliance
+    with the License.  You may obtain a copy of the License at
+
+..    http://www.apache.org/licenses/LICENSE-2.0
+
+..  Unless required by applicable law or agreed to in writing,
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+===========================================
+Python User-defined Table Functions (UDTFs)
+===========================================
+
+Spark 3.5 introduces a new type of user-defined fucntion: Python user-defined table functions (UDTFs),
+which take zero or more arguments and return a set of rows.
+
+Implementing a Python UDTF
+--------------------------
+
+.. currentmodule:: pyspark.sql.functions
+
+To implement a Python UDTF, you can implement this class:
+
+.. code-block:: python
+
+    class PythonUDTF:
+
+        def __init__(self) -> None:
+            """
+            Initialize the user-defined table function (UDTF).
+
+            This method is optional to implement and is called once when the UDTF is
+            instantiated. Use it to perform any initialization required for the UDTF.
+            """
+            ...
+
+        def eval(self, *args: Any) -> Iterator[Any]:
+            """"
+            Evaluate the function using the given input arguments.
+
+            This method is required to implement.
+
+            Args:
+                *args: Arbitrary positional arguments representing the input
+                       to the UDTF.
+
+            Yields:
+                tuple: A tuple representing a single row in the UDTF result relation.
+                       Yield thisas many times as needed to produce multiple rows.
+
+            Note:
+                - The result must be a tuple.
+                - UDTFs do not accept keyword arguments on the calling side.
+                - Use "yield" to produce one row at a time for the UDTF result relation,
+                  or use "return" to produce multiple rows for the UDTF result relation at once.
+
+            Example:
+                def eval(self, x: int, y: int):
+                    yield x + y, x - y
+            """
+            ...
+
+        def terminate(self) -> Iterator[Any]:
+            """
+            Called when the UDTF has processed all rows in a partition.
```

Comment:
We haven't really precisely defined what comprises a partition yet. Should we define it using the definitions from https://github.com/apache/spark/pull/42100 and https://github.com/apache/spark/pull/42174? Alternatively if these docs are targeting Spark 3.5 but those PRs are only going into `master`, we could simply define a partition here as either (1) just one `eval` call with the provided scalar argument(s), if any, or (2) several `eval` calls with an undefined subset of the rows from the input relation. Then we can expand it later.